### PR TITLE
Fix issue with vertical lines not visible in the elevation profile tool

### PR DESCRIPTION
**Pull request by ptitjano**
## Description

This PR reverts commit ce96dee6b517bc19a84c9394e48d6cb30e80c1f7.

This commit was introduced to gain some performance when the
`tolerance` is set by using `request.setDistanceWithin` instead of
`request.setFilterRect`. However, in that case, the providers either
use `GEOS::distance` or `GEOS::distanceWithin` which is not able to
handle the Z component of a Geometry.
For example, a purely vertical line, i.e.: `LineString(X Y Z1, X Y
Z2)` will always be to an infinite distance of any geometry according
to GEOS.

In practice, this means that a purely vertical line will never be
visible in the elevation profile when the tolerance is set. Indeed,
the provider will always discard such geometry because of the usage
of `setDistanceWithin`. This issue is fixed by always using
`setFilterRect`. This might introduce a performance penalty but this
ensures that the result is always correct.

## Update October 10

```
`GEOSPreparedDistance_r` is not able to properly handle a purely
vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))). In that
case, it always `inf`. However, `GEOSDistance_r` works.

This issue is fixed by checking if one of the geometry is vertical 3d
line by introducing `isZVerticalLine`. This function checks if a
geometry is a 3d line. If that's the case, it loops through the points
of the line. If two points do not have the same X Y corrdinates, it
means that the line is not purely vertical.

If `geom` is a vertical line, it is replaced by its first point.
It the prepared geometry is a vertical line, fallback to the non
prepared case.
```

### CODE REVIEW

**Comment by nyalldawson** [comment]
> However, in that case, the providers either
use GEOS::distance or GEOS::distanceWithin which is not able to
handle the Z component of a Geometry.
For example, a purely vertical line, i.e.: LineString(X Y Z1, X Y Z2) will always be to an infinite distance of any geometry according
to GEOS.

I think this special case should be handled in the iterator (or geometry) classes then, not specifically the elevation profile code

**Comment by ptitjano** [comment]
> > However, in that case, the providers either
> > use GEOS::distance or GEOS::distanceWithin which is not able to
> > handle the Z component of a Geometry.
> > For example, a purely vertical line, i.e.: LineString(X Y Z1, X Y Z2) will always be to an infinite distance of any geometry according
> > to GEOS.
> 
> I think this special case should be handled in the iterator (or geometry) classes then, not specifically the elevation profile code

Thanks for you message. However, I don't understand it. Do you mean that the iterator code of the providers should be changed? Or something else?

**Comment by nyalldawson** [comment]
. Do you mean that the iterator code of the providers should be changed?

I mean if the setDistanceWithin method for QgsFeatureRequest can't handle this situation, then it needs to be fixed in the iterator code. Otherwise other users of setDistanceWithin will remain broken. 

So maybe the fix needs to apply in QgsGeos::distance / QgsGeos::distanceWithin instead, eg by detecting a vertical line and using a GEOS point geometry there instead...

**Comment by ptitjano** [comment]
> . Do you mean that the iterator code of the providers should be changed?
> 
> I mean if the setDistanceWithin method for QgsFeatureRequest can't handle this situation, then it needs to be fixed in the iterator code. Otherwise other users of setDistanceWithin will remain broken.
> 
> So maybe the fix needs to apply in QgsGeos::distance / QgsGeos::distanceWithin instead, eg by detecting a vertical line and using a GEOS point geometry there instead...

Thanks for the suggestion. I have completely rewritten the PR to handle this is in `QgsGeos`.

File: src/core/geometry/qgsgeos.cpp [hunk 1]
```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
...
+{
+  // checks if the Geometry if a purely vertical 3D line LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))
+  // This is needed because QgsGeos is not able to handle this type of geometry on distance computation.
+
+  if ( geom->wkbType() != Qgis::WkbType::LineStringZ && geom->wkbType() != Qgis::WkbType::LineStringZM )
+  {
+    return false;
+  }
+
+  std::unique_ptr<QgsLineString> line( qgsgeometry_cast<QgsLineString *>( geom->clone() ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:554]: Why a clone vs just casting in place?

File: src/core/geometry/qgsgeos.cpp [hunk 2]
```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
...
+  bool isVertical = true;
+  if ( line )
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:556]:
```suggestion
  if ( const QgsLineString* line = qgsgeometry_cast<const QgsLineString *>( geom ) )
```

File: src/core/geometry/qgsgeos.cpp [hunk 3]
```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
...
+    {
+      return true;
+    }
+
+    // if the 2D part of two points of the line are different, this means
+    // that the line is not purely vertical
+    const double sqrTolerance = tolerance * tolerance;
+    for ( int iVert = nrPoints - 1, jVert = 0; jVert < nrPoints; iVert = jVert++ )
+    {
+      const QgsPoint pi = geom->vertexAt( QgsVertexId( 0, 0, iVert ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:569]: vertexAt is a slow approach to iterate through vertices -- better is something like https://github.com/qgis/QGIS/blob/master/src/core/symbology/qgssymbollayerutils.cpp#L4452 .

File: src/core/geometry/qgsgeos.cpp [hunk 4]
```diff
@@ -548,7 +587,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
     return distance;
   }
 
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  // GEOSPreparedDistance_r is not able to properly compute the distance if one
+  // of the geometries if a vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))).
+  // In that case, replace `geom` by a single point.
+  // However, GEOSDistance_r works.
+  std::unique_ptr< QgsAbstractGeometry > preparedGeom;
+  if ( mGeosPrepared && isZVerticalLine( geom ) )
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:595]: There's an issue here if geom is a multilinestring containing a single line. Best way to handle this is:

```suggestion
  if ( mGeosPrepared && isZVerticalLine( geom->simplifiedTypeRef() ) )
```

File: src/core/geometry/qgsgeos.cpp [hunk 5]
```diff
@@ -548,7 +587,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
...
+  {
+    preparedGeom.reset( geom->vertexAt( QgsVertexId( 0, 0, 0 ) ).clone() );
+  }
+  else
+  {
+    preparedGeom.reset( geom->clone() );
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:601]: This is inefficient -- it's forcing an unnecessary clone in the standard (non vertical line) case. 

Move `geos::unique_ptr otherGeosGeom'` to before this check, and avoid setting preparedGeom when we don't need it.

File: src/core/geometry/qgsgeos.cpp [hunk 6]
```diff
@@ -607,7 +660,17 @@ bool QgsGeos::distanceWithin( const QgsAbstractGeometry *geom, double maxdist, Q
     return false;
   }
 
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  std::unique_ptr< QgsAbstractGeometry > preparedGeom;
+  if ( mGeosPrepared && isZVerticalLine( geom ) )
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:664]: Same notes as above -- handle multilinestrings, avoid unwanted clone

**Comment by nyalldawson** [comment]
Thanks @ptitjano , this approach looks much better to me! Just be aware of introducing any inefficiencies in these classes (like clones which aren't 100% required), as this is all performance critical code which needs to be as optimised as possible

File: src/core/geometry/qgsgeos.cpp (hunk 1 above)
Comment (ptitjano) [reply src/core/geometry/qgsgeos.cpp]: This was la left-over from a previous iteration of the change. Fixed.

File: src/core/geometry/qgsgeos.cpp (hunk 2 above)
Comment (ptitjano) [reply src/core/geometry/qgsgeos.cpp]: Changed

File: src/core/geometry/qgsgeos.cpp (hunk 3 above)
Comment (ptitjano) [reply src/core/geometry/qgsgeos.cpp]: This should be good now.

File: src/core/geometry/qgsgeos.cpp (hunk 4 above)
Comment (ptitjano) [reply src/core/geometry/qgsgeos.cpp]: Done.

File: src/core/geometry/qgsgeos.cpp (hunk 5 above)
Comment (ptitjano) [reply src/core/geometry/qgsgeos.cpp]: Done.

**Comment by ptitjano** [comment]
> Thanks @ptitjano , this approach looks much better to me! Just be aware of introducing any inefficiencies in these classes (like clones which aren't 100% required), as this is all performance critical code which needs to be as optimised as possible

Thanks. This should be good now.

File: src/core/geometry/qgsgeos.cpp [hunk 7]
```diff
@@ -548,7 +586,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
...
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  geos::unique_ptr otherGeosGeom;
+
+  // GEOSPreparedDistance_r is not able to properly compute the distance if one
+  // of the geometries if a vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))).
+  // In that case, replace `geom` by a single point.
+  // However, GEOSDistance_r works.
+  if ( mGeosPrepared && isZVerticalLine( geom->simplifiedTypeRef() ) )
+  {
+    otherGeosGeom = asGeos( geom->vertexAt( QgsVertexId( 0, 0, 0 ) ).clone(), mPrecision );
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeos.cpp:597]: asGeos doesn't take ownership, so this leaks the cloned point

Comment (ptitjano) [reply src/core/geometry/qgsgeos.cpp]: Fixed.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**